**data accesing:**
Types of Unclean Data
There are 2 kinds of unclean data

a. Dirty Data (Data with Quality issues): Dirty data, also known as low quality data. Low quality data has content issues.

1. Duplicated data
2. Missing Data
3. Corrupt Data
4. Inaccurate Data

b. Messy Data (Data with tidiness issues): Messy data, also known as untidy data. Untidy data has structural issues.Tidy data has the following properties:

1. Each variable forms a column
2. Each observation forms a row
3. Each observational unit forms a table

In [1]:
import numpy as np
import pandas as pd

In [3]:
patients = pd.read_csv("../datasets/patients.csv")
adverse_reactions = pd.read_csv("../datasets/adverse_reactions.csv")
treatments = pd.read_csv("../datasets/treatments.csv")
treatments_cut = pd.read_csv("../datasets/treatments_cut.csv")

In [4]:
patients.head()

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
0,1,female,Zoe,Wellish,576 Brown Bear Drive,Rancho California,California,92390.0,United States,951-719-9170ZoeWellish@superrito.com,7/10/1976,121.7,66,19.6
1,2,female,Pamela,Hill,2370 University Hill Road,Armstrong,Illinois,61812.0,United States,PamelaSHill@cuvox.de+1 (217) 569-3204,4/3/1967,118.8,66,19.2
2,3,male,Jae,Debord,1493 Poling Farm Road,York,Nebraska,68467.0,United States,402-363-6804JaeMDebord@gustr.com,2/19/1980,177.8,71,24.8
3,4,male,Liêm,Phan,2335 Webster Street,Woodbridge,NJ,7095.0,United States,PhanBaLiem@jourrapide.com+1 (732) 636-8246,7/26/1951,220.9,70,31.7
4,5,male,Tim,Neudorf,1428 Turkey Pen Lane,Dothan,AL,36303.0,United States,334-515-7487TimNeudorf@cuvox.de,2/18/1928,192.3,27,26.1


In [5]:
adverse_reactions.head()

,given_name,surname,adverse_reaction
0,berta,napolitani,injection site discomfort
1,lena,baer,hypoglycemia
2,joseph,day,hypoglycemia
3,flavia,fiorentino,cough
4,manouck,wubbels,throat irritation


In [6]:
treatments.head()

,given_name,surname,auralin,novodra,hba1c_start,hba1c_end,hba1c_change
0,veronika,jindrová,41u - 48u,-,7.63,7.20,NaN
1,elliot,richardson,-,40u - 45u,7.56,7.09,0.97
2,yukitaka,takenaka,-,39u - 36u,7.68,7.25,NaN
3,skye,gormanston,33u - 36u,-,7.97,7.62,0.35
4,alissa,montez,-,33u - 29u,7.78,7.46,0.32


In [7]:
treatments_cut.head()

,given_name,surname,auralin,novodra,hba1c_start,hba1c_end,hba1c_change
0,jožka,resanovič,22u - 30u,-,7.56,7.22,0.34
1,inunnguaq,heilmann,57u - 67u,-,7.85,7.45,NaN
2,alwin,svensson,36u - 39u,-,7.78,7.34,NaN
3,thể,lương,-,61u - 64u,7.64,7.22,0.92
4,amanda,ribeiro,36u - 44u,-,7.85,7.47,0.38


2 types of accessing data:

a. Manual

b. programming

### step 1: Write a summary for your data

This is a dataset about 500 patients of which 350 patients participated in a clinical trial. None of the patients were using Novodra (a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before. All were experiencing elevated HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After 4 weeks, which isn’t enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:

- 175 patients switched to Auralin for 24 weeks
- 175 patients continued using Novodra for 24 weeks

Data about patients feeling some adverse effects is also recorded.

### step 2: Write Column descriptions

#### **Table** -> `patients`:

- `patient_id`: the unique identifier for each patient in the Master Patient Index (i.e. patient database) of the pharmaceutical company that is producing Auralin
- `assigned_sex`: the assigned sex of each patient at birth (male or female)
- `given_name`: the given name (i.e. first name) of each patient
- `surname`: the surname (i.e. last name) of each patient
- `address`: the main address for each patient
- `city`: the corresponding city for the main address of each patient
- `state`: the corresponding state for the main address of each patient
- `zip_code`: the corresponding zip code for the main address of each patient
- `country`: the corresponding country for the main address of each patient (all United states for this clinical trial)
- `contact`: phone number and email information for each patient
- `birthdate`: the date of birth of each patient (month/day/year). The inclusion criteria for this clinical trial is age >= 18 (there is no maximum age because diabetes is a growing problem among the elderly population)
- `weight`: the weight of each patient in pounds (lbs)
- `height`: the height of each patient in inches (in)
- `bmi`: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m2 where kg is a person's weight in kilograms and m2 is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. The inclusion criteria for this clinical trial is 16 >= BMI >= 38.

#### **Table** -> `treatments` and `treatment_cut`:

- `given_name`: the given name of each patient in the Master Patient Index that took part in the clinical trial
- `surname`: the surname of each patient in the Master Patient Index that took part in the clinical trial
- `auralin`: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash) and the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash). Both are measured in units (shortform 'u'), which is the international unit of measurement and the standard measurement for insulin.
- `novodra`: same as above, except for patients that continued treatment with Novodra
- `hba1c_start`: the patient's HbA1c level at the beginning of the first week of treatment. HbA1c stands for Hemoglobin A1c. The HbA1c test measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.
- `hba1c_end`: the patient's HbA1c level at the end of the last week of treatment
- `hba1c_change`: the change in the patient's HbA1c level from the start of treatment to the end, i.e., hba1c_start - hba1c_end. For Auralin to be deemed effective, it must be "noninferior" to Novodra, the current standard for insulin. This "noninferiority" is statistically defined as the upper bound of the 95% confidence interval being less than 0.4% for the difference between the mean HbA1c changes for Novodra and Auralin (i.e. Novodra minus Auralin).

#### **Table** -> `adverse_reactions`

- `given_name`: the given name of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- `surname`: the surname of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- `adverse_reaction`: the adverse reaction reported by the patient

### 3. Add any additional information

Additional useful information:

- Insulin resistance varies person to person, which is why both starting median daily dose and ending median daily dose are required, i.e., to calculate change in dose.
- It is important to test drugs and medical products in the people they are meant to help. People of different age, race, sex, and ethnic group must be included in clinical trials. This diversity is reflected in the patients table.

### Main steps start now:   Types of Assessment
There are 2 types of assessment styles

- `Manual` - Looking through the data manually in google sheets
- `Programmatic` - By using pandas functions such as info(), describe() or sample()

In [425]:
#Manual : export data for manual assessment:

with pd.ExcelWriter("Clinical_trials.xlsx") as writer:    #pd.excelwriter is used when we have multiple sheets in excel file
    patients.to_excel(writer, sheet_name="patients", index=False)
    adverse_reactions.to_excel(writer, sheet_name="adverse_reactions", index=False)
    treatments.to_excel(writer, sheet_name="treatments", index=False)
    treatments_cut.to_excel(writer, sheet_name="treatments_cut", index=False)

In [426]:
#now file that you have created above you have to download it from file explorer in "users/adity/favorite":  path and then open in excel

**1. Manually: Write Issues with the dataset: yeh aapko data mein dekhke hi pata lag skta hai**

a. Dirty Data

Table - Patients

using manual:
- patient_id = 9 has misspelled name 'Dsvid' instead of David `accuracy`
- state col sometimes contain full name and some times abbrivietation `consistency`
- zip code col has entries with 4 digit `validity`

using programming:
- data missing for 12 patients in address,city, state,zip_code  ,country,   contact `completion`
- incorrect data type assigned to sex, zip code, birthdate `validity`
- duplicate entries by the name of John Doe `accuracy`
- one patient has weight = 48 pounds `accuracy`
- one patient has height = 27 inches `accuracy`


Table - Treatments & Treatments_cut

- given_name and surname col is is all lower case `consistency`
- remove u from Auralin and Novadra cols `validity`
- '-' in novadra and Auralin col treated as nan `validity`
- missing values in hba1c_change col `completion`
- 1 duplicate entry by the name Joseph day `accuracy`
- in hba1c_change 9 instead of 4 `accuracy`

Table - Adverse_reactions

- given_name and surname are all in lower case `consistency`

  
b. Messy Data

Table - Patients

- contact col contains both phone and email

Table - Treatments & Treatments_cut

- Auralin and Novadra col should be split into 2 cols start and end dose
- merge both the tables

Table - Adverse_reactions

- This table should not exist independently

**2. Automatic Assessment: neeche functions chalaake pata krna pdega**

- head and tail
- sample
- info
- isnull
- duplicated
- describe

In [427]:
patients.sample(5)  # random 5 rows for better knowledge

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
301,302,female,Onyekachukwu,Obinna,2970 Forest Avenue,New York,NY,10004.0,United States,OnyekachukwuObinna@teleworm.us646-982-6609,1/24/1997,154.7,65,25.7
299,300,female,Linda,Lundy,3964 Walnut Avenue,Bovill,ID,83823.0,United States,208-826-1678LindaJLundy@gustr.com,1/29/1969,192.5,69,28.4
20,21,female,Sofia,Karlsen,2931 Romano Street,Whitman,MA,2382.0,United States,SofiaTKarlsen@teleworm.us1 781 447 1763,9/24/1934,153.1,66,24.7
101,102,male,Oles,Zhdanov,1731 Chandler Drive,Fordland,MO,65652.0,United States,OlesZhdanov@fleckens.hu417-738-5084,10/25/1945,226.2,72,30.7
46,47,male,Javier,Moquin,1808 Stuart Street,Pittsburgh,PA,15201.0,United States,724-419-3583JavierMoquin@rhyta.com,8/7/1941,163.9,72,22.2


In [428]:
patients.info()    
#now you know that total 503 entries are there but in city, state, zip_code, country and contact there are 491 entries so 12 missing entries

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [429]:
#to vo missing data hum dekh bhi skte hai like address column mein missing values dekhne ho to:

patients["address"].isnull()  # to yeh boolean de dega 

0      False
1      False
2      False
3      False
4      False
       ...  
498    False
499    False
500    False
501    False
502    False
Name: address, Length: 503, dtype: bool

In [430]:
patients[patients["address"].isnull()]   #but agar isi data mein dobara daalde to vo value bhbi miljaayegi , good part is ki baaki column ke bhi yhi rows mein missing values hain

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
209,210,female,Lalita,Eldarkhanov,NaN,NaN,NaN,NaN,NaN,NaN,8/14/1950,143.4,62,26.2
219,220,male,Mỹ,Quynh,NaN,NaN,NaN,NaN,NaN,NaN,4/9/1978,237.8,69,35.1
230,231,female,Elisabeth,Knudsen,NaN,NaN,NaN,NaN,NaN,NaN,9/23/1976,165.9,63,29.4
234,235,female,Martina,Tománková,NaN,NaN,NaN,NaN,NaN,NaN,4/7/1936,199.5,65,33.2
242,243,male,John,O'Brian,NaN,NaN,NaN,NaN,NaN,NaN,2/25/1957,205.3,74,26.4
249,250,male,Benjamin,Mehler,NaN,NaN,NaN,NaN,NaN,NaN,10/30/1951,146.5,69,21.6
257,258,male,Jin,Kung,NaN,NaN,NaN,NaN,NaN,NaN,5/17/1995,231.7,69,34.2
264,265,female,Wafiyyah,Asfour,NaN,NaN,NaN,NaN,NaN,NaN,11/3/1989,158.6,63,28.1
269,270,female,Flavia,Fiorentino,NaN,NaN,NaN,NaN,NaN,NaN,10/9/1937,175.2,61,33.1
278,279,female,Generosa,Cabán,NaN,NaN,NaN,NaN,NaN,NaN,12/16/1962,124.3,69,18.4


In [431]:
treatments.info()  #  hba1c_change m,issing values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [432]:
treatments_cut.info()   #  hba1c_change m,issing values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  42 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


In [433]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 948.0+ bytes


**Now check if any duplicate value is there**

In [434]:
patients.duplicated().sum()

np.int64(0)

In [435]:
treatments.duplicated().sum()

np.int64(1)

In [436]:
treatments[treatments.duplicated()]   # isi pe daalke vo missing value bhi dekhlo

,given_name,surname,auralin,novodra,hba1c_start,hba1c_end,hba1c_change
136,joseph,day,29u - 36u,-,7.7,7.19,NaN


In [437]:
treatments_cut.duplicated().sum()

np.int64(0)

In [438]:
adverse_reactions.duplicated().sum()

np.int64(0)

In [439]:
patients.describe()

,patient_id,zip_code,weight,height,bmi
count,503.000000,491.000000,503.000000,503.000000,503.000000
mean,252.000000,49084.118126,173.434990,66.634195,27.483897
std,145.347859,30265.807442,33.916741,4.411297,5.276438
min,1.000000,1002.000000,48.800000,27.000000,17.100000
25%,126.500000,21920.500000,149.300000,63.000000,23.300000
50%,252.000000,48057.000000,175.300000,67.000000,27.200000
75%,377.500000,75679.000000,199.500000,70.000000,31.750000
max,503.000000,99701.000000,255.900000,79.000000,37.700000


In [440]:
patients[patients["weight"] == 48.800000]   # yeh wrong lag rha hai as 21 kg nhi ho skta women jiska birth 1938 mein hua ho

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
210,211,female,Camilla,Zaitseva,4689 Briarhill Lane,Wooster,OH,44691.0,United States,330-202-2145CamillaZaitseva@superrito.com,11/26/1938,48.8,63,19.1


**Note - Assessing Data is an Iterative Process**

1. Data Quality Dimensions:

Completeness -> is data missing?

Validity -> is data invalid -> negative height -> duplicate patient id

Accuracy -> data is valid but not accurate -> weight -> 1kg

Consistency -> both valid and accurate but written differently -> New Youk and NY

2. Order of severity:

Completeness <- Validity <- Accuracy <- Consistency

3. Data Cleaning Order:

Quality -> Completeness

Tidiness

Quality -> Validity

Quality -> Accuracy

Quality -> Consistency


4. Steps involved in Data cleaning

Define

Code

Test

**note: Always make sure to create a copy of your pandas dataframe before you start the cleaning process**

In [441]:
patients_df = patients.copy()
treatments_df = treatments.copy()
treatments_cut_df = treatments_cut.copy()
adverse_reactions_df = adverse_reactions.copy()

**now start cleaning according to Data Quality Dimensions:**

**step 1. completeness: pehle jitne bhi completion vaali problem hai use solve krenge**

In [442]:
patients_df.info()  #fill missing values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [443]:
patients_df[patients_df["address"].isnull()]

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
209,210,female,Lalita,Eldarkhanov,NaN,NaN,NaN,NaN,NaN,NaN,8/14/1950,143.4,62,26.2
219,220,male,Mỹ,Quynh,NaN,NaN,NaN,NaN,NaN,NaN,4/9/1978,237.8,69,35.1
230,231,female,Elisabeth,Knudsen,NaN,NaN,NaN,NaN,NaN,NaN,9/23/1976,165.9,63,29.4
234,235,female,Martina,Tománková,NaN,NaN,NaN,NaN,NaN,NaN,4/7/1936,199.5,65,33.2
242,243,male,John,O'Brian,NaN,NaN,NaN,NaN,NaN,NaN,2/25/1957,205.3,74,26.4
249,250,male,Benjamin,Mehler,NaN,NaN,NaN,NaN,NaN,NaN,10/30/1951,146.5,69,21.6
257,258,male,Jin,Kung,NaN,NaN,NaN,NaN,NaN,NaN,5/17/1995,231.7,69,34.2
264,265,female,Wafiyyah,Asfour,NaN,NaN,NaN,NaN,NaN,NaN,11/3/1989,158.6,63,28.1
269,270,female,Flavia,Fiorentino,NaN,NaN,NaN,NaN,NaN,NaN,10/9/1937,175.2,61,33.1
278,279,female,Generosa,Cabán,NaN,NaN,NaN,NaN,NaN,NaN,12/16/1962,124.3,69,18.4


In [444]:
patients_df.fillna("Miss Data", inplace = True)

C:\Users\adity\AppData\Local\Temp\ipykernel_12080\1281361533.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'Miss Data' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  patients_df.fillna("Miss Data", inplace = True)


In [445]:
patients_df.info()   #now all columns are filled

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       503 non-null    object 
 5   city          503 non-null    object 
 6   state         503 non-null    object 
 7   zip_code      503 non-null    object 
 8   country       503 non-null    object 
 9   contact       503 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(2), int64(2), object(10)
memory usage: 55.1+ KB


In [446]:
treatments_df.head()

,given_name,surname,auralin,novodra,hba1c_start,hba1c_end,hba1c_change
0,veronika,jindrová,41u - 48u,-,7.63,7.20,NaN
1,elliot,richardson,-,40u - 45u,7.56,7.09,0.97
2,yukitaka,takenaka,-,39u - 36u,7.68,7.25,NaN
3,skye,gormanston,33u - 36u,-,7.97,7.62,0.35
4,alissa,montez,-,33u - 29u,7.78,7.46,0.32


In [447]:
treatments_df["hba1c_change"] = treatments_df["hba1c_start"] - treatments_df["hba1c_end"]
treatments_df.head()      #now values are correct

,given_name,surname,auralin,novodra,hba1c_start,hba1c_end,hba1c_change
0,veronika,jindrová,41u - 48u,-,7.63,7.20,0.43
1,elliot,richardson,-,40u - 45u,7.56,7.09,0.47
2,yukitaka,takenaka,-,39u - 36u,7.68,7.25,0.43
3,skye,gormanston,33u - 36u,-,7.97,7.62,0.35
4,alissa,montez,-,33u - 29u,7.78,7.46,0.32


In [448]:
treatments_cut_df["hba1c_change"] = treatments_cut_df["hba1c_start"] - treatments_cut_df["hba1c_end"]
treatments_cut_df.head()      #now values are correct

,given_name,surname,auralin,novodra,hba1c_start,hba1c_end,hba1c_change
0,jožka,resanovič,22u - 30u,-,7.56,7.22,0.34
1,inunnguaq,heilmann,57u - 67u,-,7.85,7.45,0.40
2,alwin,svensson,36u - 39u,-,7.78,7.34,0.44
3,thể,lương,-,61u - 64u,7.64,7.22,0.42
4,amanda,ribeiro,36u - 44u,-,7.85,7.47,0.38


**step2: now tidiness so move to messy data:**

In [449]:
patients.head()    #in contact column phone no and email must be different coilumn

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
0,1,female,Zoe,Wellish,576 Brown Bear Drive,Rancho California,California,92390.0,United States,951-719-9170ZoeWellish@superrito.com,7/10/1976,121.7,66,19.6
1,2,female,Pamela,Hill,2370 University Hill Road,Armstrong,Illinois,61812.0,United States,PamelaSHill@cuvox.de+1 (217) 569-3204,4/3/1967,118.8,66,19.2
2,3,male,Jae,Debord,1493 Poling Farm Road,York,Nebraska,68467.0,United States,402-363-6804JaeMDebord@gustr.com,2/19/1980,177.8,71,24.8
3,4,male,Liêm,Phan,2335 Webster Street,Woodbridge,NJ,7095.0,United States,PhanBaLiem@jourrapide.com+1 (732) 636-8246,7/26/1951,220.9,70,31.7
4,5,male,Tim,Neudorf,1428 Turkey Pen Lane,Dothan,AL,36303.0,United States,334-515-7487TimNeudorf@cuvox.de,2/18/1928,192.3,27,26.1


In [450]:
#Now we use regex (regular expression to seperate it):
# Extract email
# Extract only the email address (ignore any digits/hyphens before it)
patients["email"] = patients["contact"].str.extract(
    r'([A-Za-z][A-Za-z0-9._-]*@[A-Za-z0-9.-]+\.\w+)'
)

# Extract phone number (works for both formats)
patients["phone_no"] = patients["contact"].str.extract(
    r'(\+?\d?\s*\(?\d{3}\)?[- ]\d{3}-\d{4})'
)

# Drop original column
patients.drop(columns="contact", inplace=True)

patients.head()

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,birthdate,weight,height,bmi,email,phone_no
0,1,female,Zoe,Wellish,576 Brown Bear Drive,Rancho California,California,92390.0,United States,7/10/1976,121.7,66,19.6,ZoeWellish@superrito.com,951-719-9170
1,2,female,Pamela,Hill,2370 University Hill Road,Armstrong,Illinois,61812.0,United States,4/3/1967,118.8,66,19.2,PamelaSHill@cuvox.de,+1 (217) 569-3204
2,3,male,Jae,Debord,1493 Poling Farm Road,York,Nebraska,68467.0,United States,2/19/1980,177.8,71,24.8,JaeMDebord@gustr.com,402-363-6804
3,4,male,Liêm,Phan,2335 Webster Street,Woodbridge,NJ,7095.0,United States,7/26/1951,220.9,70,31.7,PhanBaLiem@jourrapide.com,+1 (732) 636-8246
4,5,male,Tim,Neudorf,1428 Turkey Pen Lane,Dothan,AL,36303.0,United States,2/18/1928,192.3,27,26.1,TimNeudorf@cuvox.de,334-515-7487


**Explanation:**

**Step 1**: patients["email"] = patients["contact"].str.extract(...) = Go to the contact column.Find only the email address.Store it in a new column called email.

Think of it as saying: "Ignore everything else and copy only the email."

**Step 2**: r'([A-Za-z][A-Za-z0-9._-]*@[A-Za-z0-9.-]+\.\w+)' = This is the pattern used to identify an email. Let's break it into small pieces.

Part 1: [A-Za-z] Means: The email must start with a letter.

Examples:

✅

Zoe, Jae, Tim, Pamela

❌

951, 402, 334

This is why: 951-719-9170ZoeWellish@superrito.com starts matching from Z instead of 9

Part 2: [A-Za-z0-9._-]*: Means: After the first letter, the email can contain letters,numbers,dot .,underscore _ and hyphen -

The * means: Zero or more characters. So it matches

ZoeWellish, Jae123, john.doe, abc_xyz

Part 3: @: Simply matches the @ symbol.

Part 4: [A-Za-z0-9.-]+: Matches the website name. Examples: gmail, cuvox, gustr and superrito

Part 5: \.: Means Match the actual dot .

Part 6: \w+: Matches com, de, org, in So together: [A-Za-z][A-Za-z0-9._-]*@[A-Za-z0-9.-]+\.\w+

finds: ZoeWellish@superrito.com, PamelaSHill@cuvox.de, JaeMDebord@gustr.com and ignores

951-719-9170, 402-363-6804

**Step 3**: patients["phone_no"] = patients["contact"].str.extract(...) = Now we go to the same contact column. This time we ignore the email and search only for the phone number.

**Step 4**: r'(\+?\d?\s*\(?\d{3}\)?[- ]\d{3}-\d{4})' = This is the phone number pattern. You don't have to memorize it. Just know it recognizes phone numbers like 951-719-9170, 402-363-6804, +1 (217) 569-3204, +1 (732) 636-8246 and copies them into the phone_no column.

**Step 5**: patients.drop(columns="contact", inplace=True) = Now we already have email and phone_no. So the old contact column is no longer needed. This line simply deletes it.

Regex Cheat Sheet:

Symbol:     Meaning

1. r'' :    Raw string
2. ()  :    Capture the match
3. []  :    Match one character from the list
4. \w  :    Letter, digit, or underscore
5. \d  :    Digit (0–9)
6. \. :     Literal dot (.)
7. \s  :    Space character
8. "+" :     One or more
9. "*"  :     Zero or more
10. ? :     Optional (zero or one)
11. {3}:    Exactly 3 times
12. {4}:	   Exactly 4 times

In [451]:
#now conytact two tables:

treatments_df = pd.concat([treatments_df, treatments_cut_df])
treatments_df

,given_name,surname,auralin,novodra,hba1c_start,hba1c_end,hba1c_change
0,veronika,jindrová,41u - 48u,-,7.63,7.20,0.43
1,elliot,richardson,-,40u - 45u,7.56,7.09,0.47
2,yukitaka,takenaka,-,39u - 36u,7.68,7.25,0.43
3,skye,gormanston,33u - 36u,-,7.97,7.62,0.35
4,alissa,montez,-,33u - 29u,7.78,7.46,0.32
...,...,...,...,...,...,...,...
65,rovzan,kishiev,32u - 37u,-,7.75,7.41,0.34
66,jakob,jakobsen,-,28u - 26u,7.96,7.51,0.45
67,bernd,schneider,48u - 56u,-,7.74,7.44,0.30
68,berta,napolitani,-,42u - 44u,7.68,7.21,0.47


In [452]:
#now auralin and novodra column shoulb be as this is dosage column: so we use melt() function:

treatments_df = treatments_df.melt(id_vars= ["given_name","surname","hba1c_start", "hba1c_end", "hba1c_change"], var_name ="dose_type", value_name = "dose_range")
treatments_df

,given_name,surname,hba1c_start,hba1c_end,hba1c_change,dose_type,dose_range
0,veronika,jindrová,7.63,7.20,0.43,auralin,41u - 48u
1,elliot,richardson,7.56,7.09,0.47,auralin,-
2,yukitaka,takenaka,7.68,7.25,0.43,auralin,-
3,skye,gormanston,7.97,7.62,0.35,auralin,33u - 36u
4,alissa,montez,7.78,7.46,0.32,auralin,-
...,...,...,...,...,...,...,...
695,rovzan,kishiev,7.75,7.41,0.34,novodra,-
696,jakob,jakobsen,7.96,7.51,0.45,novodra,28u - 26u
697,bernd,schneider,7.74,7.44,0.30,novodra,-
698,berta,napolitani,7.68,7.21,0.47,novodra,42u - 44u


**now above table have 700 entries but hamein 350 hi chahiye so hum hyphen(-) values hata denge kyunki koi bhi patient dono to nhi le rha** 

In [453]:
treatments_df["dose_range"].str.strip() !="-"

0       True
1      False
2      False
3       True
4      False
       ...  
695    False
696     True
697    False
698     True
699    False
Name: dose_range, Length: 700, dtype: bool

In [454]:
treatments_df = treatments_df[treatments_df["dose_range"].str.strip() !="-"]   #isi mein store krdo
treatments_df

,given_name,surname,hba1c_start,hba1c_end,hba1c_change,dose_type,dose_range
0,veronika,jindrová,7.63,7.20,0.43,auralin,41u - 48u
3,skye,gormanston,7.97,7.62,0.35,auralin,33u - 36u
6,sophia,haugen,7.65,7.27,0.38,auralin,37u - 42u
7,eddie,archer,7.89,7.55,0.34,auralin,31u - 38u
9,asia,woźniak,7.76,7.37,0.39,auralin,30u - 36u
...,...,...,...,...,...,...,...
688,christopher,woodward,7.51,7.06,0.45,novodra,55u - 51u
690,maret,sultygov,7.67,7.30,0.37,novodra,26u - 23u
694,lixue,hsueh,9.21,8.80,0.41,novodra,22u - 23u
696,jakob,jakobsen,7.96,7.51,0.45,novodra,28u - 26u


**in dose range there are 2 values but single value honi chahiye to use remove krdenge**

In [455]:
treatments_df["dose_start"] = treatments_df["dose_range"].str.split("-").str.get(0)                                   
treatments_df["dose_end"] = treatments_df["dose_range"].str.split("-").str.get(1) 

C:\Users\adity\AppData\Local\Temp\ipykernel_12080\3885804675.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  treatments_df["dose_start"] = treatments_df["dose_range"].str.split("-").str.get(0)
C:\Users\adity\AppData\Local\Temp\ipykernel_12080\3885804675.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  treatments_df["dose_end"] = treatments_df["dose_range"].str.split("-").str.get(1)


In [456]:
treatments_df

,given_name,surname,hba1c_start,hba1c_end,hba1c_change,dose_type,dose_range,dose_start,dose_end
0,veronika,jindrová,7.63,7.20,0.43,auralin,41u - 48u,41u,48u
3,skye,gormanston,7.97,7.62,0.35,auralin,33u - 36u,33u,36u
6,sophia,haugen,7.65,7.27,0.38,auralin,37u - 42u,37u,42u
7,eddie,archer,7.89,7.55,0.34,auralin,31u - 38u,31u,38u
9,asia,woźniak,7.76,7.37,0.39,auralin,30u - 36u,30u,36u
...,...,...,...,...,...,...,...,...,...
688,christopher,woodward,7.51,7.06,0.45,novodra,55u - 51u,55u,51u
690,maret,sultygov,7.67,7.30,0.37,novodra,26u - 23u,26u,23u
694,lixue,hsueh,9.21,8.80,0.41,novodra,22u - 23u,22u,23u
696,jakob,jakobsen,7.96,7.51,0.45,novodra,28u - 26u,28u,26u


In [457]:
treatments_df.drop(columns = "dose_range", inplace = True)

C:\Users\adity\AppData\Local\Temp\ipykernel_12080\3445101433.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  treatments_df.drop(columns = "dose_range", inplace = True)


In [458]:
treatments_df

,given_name,surname,hba1c_start,hba1c_end,hba1c_change,dose_type,dose_start,dose_end
0,veronika,jindrová,7.63,7.20,0.43,auralin,41u,48u
3,skye,gormanston,7.97,7.62,0.35,auralin,33u,36u
6,sophia,haugen,7.65,7.27,0.38,auralin,37u,42u
7,eddie,archer,7.89,7.55,0.34,auralin,31u,38u
9,asia,woźniak,7.76,7.37,0.39,auralin,30u,36u
...,...,...,...,...,...,...,...,...
688,christopher,woodward,7.51,7.06,0.45,novodra,55u,51u
690,maret,sultygov,7.67,7.30,0.37,novodra,26u,23u
694,lixue,hsueh,9.21,8.80,0.41,novodra,22u,23u
696,jakob,jakobsen,7.96,7.51,0.45,novodra,28u,26u


In [459]:
treatments_df["dose_start"] = treatments_df["dose_start"].str.replace("u","")                                   
treatments_df["dose_end"] = treatments_df["dose_end"].str.replace("u","")

C:\Users\adity\AppData\Local\Temp\ipykernel_12080\2942638308.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  treatments_df["dose_start"] = treatments_df["dose_start"].str.replace("u","")
C:\Users\adity\AppData\Local\Temp\ipykernel_12080\2942638308.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  treatments_df["dose_end"] = treatments_df["dose_end"].str.replace("u","")


In [460]:
treatments_df

,given_name,surname,hba1c_start,hba1c_end,hba1c_change,dose_type,dose_start,dose_end
0,veronika,jindrová,7.63,7.20,0.43,auralin,41,48
3,skye,gormanston,7.97,7.62,0.35,auralin,33,36
6,sophia,haugen,7.65,7.27,0.38,auralin,37,42
7,eddie,archer,7.89,7.55,0.34,auralin,31,38
9,asia,woźniak,7.76,7.37,0.39,auralin,30,36
...,...,...,...,...,...,...,...,...
688,christopher,woodward,7.51,7.06,0.45,novodra,55,51
690,maret,sultygov,7.67,7.30,0.37,novodra,26,23
694,lixue,hsueh,9.21,8.80,0.41,novodra,22,23
696,jakob,jakobsen,7.96,7.51,0.45,novodra,28,26


In [461]:
treatments_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 350 entries, 0 to 698
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   hba1c_start   350 non-null    float64
 3   hba1c_end     350 non-null    float64
 4   hba1c_change  350 non-null    float64
 5   dose_type     350 non-null    object 
 6   dose_start    350 non-null    object 
 7   dose_end      350 non-null    object 
dtypes: float64(3), object(5)
memory usage: 24.6+ KB


In [462]:
# as above column dose start and end type should be int not object

treatments_df["dose_start"] = treatments_df["dose_start"].astype("int")                                   
treatments_df["dose_end"] = treatments_df["dose_end"].astype("int") 

C:\Users\adity\AppData\Local\Temp\ipykernel_12080\2617911602.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  treatments_df["dose_start"] = treatments_df["dose_start"].astype("int")
C:\Users\adity\AppData\Local\Temp\ipykernel_12080\2617911602.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  treatments_df["dose_end"] = treatments_df["dose_end"].astype("int")


In [463]:
treatments_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 350 entries, 0 to 698
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   hba1c_start   350 non-null    float64
 3   hba1c_end     350 non-null    float64
 4   hba1c_change  350 non-null    float64
 5   dose_type     350 non-null    object 
 6   dose_start    350 non-null    int64  
 7   dose_end      350 non-null    int64  
dtypes: float64(3), int64(2), object(3)
memory usage: 24.6+ KB


In [466]:
adverse_reactions_df  # we have to merge this with treatments table as both have 2 similar column name i.e. given_name and surname

,given_name,surname,adverse_reaction
0,berta,napolitani,injection site discomfort
1,lena,baer,hypoglycemia
2,joseph,day,hypoglycemia
3,flavia,fiorentino,cough
4,manouck,wubbels,throat irritation
5,jasmine,sykes,hypoglycemia
6,louise,johnson,hypoglycemia
7,albinca,komavec,hypoglycemia
8,noe,aranda,hypoglycemia
9,sofia,hermansen,injection site discomfort


In [468]:
treatments_df.merge(adverse_reactions_df, how = "left", on = ["given_name", "surname"])   # left join kr rhe hai kunki left (treatments_df) vaala tabel hamein poora chaahiye tha

,given_name,surname,hba1c_start,hba1c_end,hba1c_change,dose_type,dose_start,dose_end,adverse_reaction
0,veronika,jindrová,7.63,7.20,0.43,auralin,41,48,NaN
1,skye,gormanston,7.97,7.62,0.35,auralin,33,36,NaN
2,sophia,haugen,7.65,7.27,0.38,auralin,37,42,NaN
3,eddie,archer,7.89,7.55,0.34,auralin,31,38,NaN
4,asia,woźniak,7.76,7.37,0.39,auralin,30,36,NaN
...,...,...,...,...,...,...,...,...,...
345,christopher,woodward,7.51,7.06,0.45,novodra,55,51,nausea
346,maret,sultygov,7.67,7.30,0.37,novodra,26,23,NaN
347,lixue,hsueh,9.21,8.80,0.41,novodra,22,23,injection site discomfort
348,jakob,jakobsen,7.96,7.51,0.45,novodra,28,26,hypoglycemia
